qiling/core.py

```diff
@@ -670,7 +688,46 @@ def enable_lib_patch(self):
                 self.mem.write(self.mem.get_lib_base(filename) + addr, code)
             except:
                 raise RuntimeError("Fail to patch %s at address 0x%x" % (filename, addr))
-    
+
+    def _init_stop_guard(self):
+        if not self.stop_options.any:
+            return
+
+        # Allocate a guard page, we need this in both cases
+        # On a negative stack pointer, we still need a return address (otherwise we end up at 0)
+        # Make sure it is not close to the heap (PE), otherwise the heap cannot grow
+        self.EXIT_TRAP_ADDR = self.mem.find_free_space(0x1000, min_addr=0x9000000, alignment=0x10)
+        self.mem.map(self.EXIT_TRAP_ADDR, 0x1000, info='[Stop guard]')
+
+        # Stop on a negative stack pointer
+        if self.stop_options.stackpointer:
+            def _check_sp(ql, address, size):
+                if not ql.loader.skip_exit_check:
+                    sp = ql._initial_sp - ql.reg.arch_sp
+                    if sp < 0:
+                        logging.info('Process returned from entrypoint (stackpointer)!')
+                        ql.emu_stop()
+
+            self.hook_code(_check_sp)
+
+        # Stop when running to exit trap address
+        if self.stop_options.exit_trap:
+            def _exit_trap(ql):
+                logging.info('Process returned from entrypoint (exit_trap)!')
+                ql.emu_stop()
+
+            self.hook_address(_exit_trap, self.EXIT_TRAP_ADDR)
+
+    def write_exit_trap(self):
+        self._initial_sp = self.reg.arch_sp
```

Comment:
Maybe this member should be put to `ql.arch` or `ql.reg`?

Reply:
It is not really a register or anything, it is more of a sentinel we need to keep track of to abort execution.
I could move it if you really want to, but in my opinion where it is now is a better place.